### ISSUE TITLE
Suggestion: Add redirection syntax that allows in-memory collection of stderr lines output by external utilities, akin to the -ErrorVariable common parameter

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#4332]
<sup>This was first proposed as part of a lengthy exchange in #3996, but I thought it deserved its own post.</sup>

**tl;dr**

In order to collected stderr output (separately) from an external-utility call in memory, I suggest supporting the following **new redirection syntax that allows capturing stderr lines in a _variable_,  akin to the common `-ErrorVariable` parameter for _cmdlets_** (the sample `cmd` command is designed to produce both stdout and stderr output):

```powershell
# Wishful thinking: Collect stderr lines in given variable $stderrOutput via 2>&
$stdoutOutput = cmd /c ver '&' dir \nosuch 2>&stderrOutput 
```

Instead of what you must currently do:

```powershell
$stderrFile = New-TemporaryFile
$stdoutOutput = cmd /c ver '&' dir \nosuch 2>$stderrFile
$stderrOutput = Get-Content $stdErrFile
Remove-Item $stderrFile
```

---

The current behavior is to pass stderr output generated by external utility calls (e.g., to `git`) through to the console - stderr output is _not_ recorded in `$Error`, which makes sense, given that many utilities write much more than just _error_ information to stderr, and that the presence of stderr input doesn't imply actual errors.

Sometimes it is necessary to inspect stderr output, however, which is currently not easily accomplished:

You can _redirect_ stderr output by redirecting PowerShell's error stream: `2>...`, but you're faced with 2 options, neither of which is convenient:

* Either: redirect stderr output to a _file_ with `2>filename`

* Or: redirect stderr output into the success stream with `2>&1` to produce a _combined_ stream from which you can later filter out the stderr lines by _type_ (`[System.Management.Automation.ErrorRecord]`).

Therefore, I propose a new redirection syntax that allows capturing stderr lines in a _variable_,  akin to the common `-ErrorVariable` parameter for _cmdlets_, so that:

* `foo.exe 2>&errs` would be the external-utility-call equivalent of a (fictitious) `Invoke-Foo -ErrorVariable errs 2>$null` call, with errors getting collected in _variable_ `$errs` in both cases.

* Additionally, to allow stderr output to _also be passed through_ while being collected:  
`foo.exe 2>&|errs` could be the equivalent of `Invoke-Foo -ErrorVariable errs`

The suggestion is to use `>&varName` syntax, because `&` is already established as having into-a-different-target(-stream) semantics, yet currently only a _digit_ is supported after the `&` (the index of a different PS stream).  
(There is still ambiguity, given that `$1` is a valid variable name, but I think that wouldn't be a real-world concern.)

Also, given that `-ErrorVariable` supports a `+` prefix to the variable name to indicate that the existing variable value should be _appended_ to, this new syntax should support that too.

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.4
```

**Comment by vexx32** [comment]
Just one question... is this generalizable to all PSProviders/PSDrives that support the Content API now?

Because IMO... it should be. 🙂

**Comment by billkindle** [comment]
I think this has some potential to be very useful especially on Linux. There’s been a few times where I’ve needed to capture standard output, didn’t care to log it, But needed to perform certain actions based on what type of output it was.

**Comment by mklement0** [comment]
Thanks, @JamesWTruher - I like the idea, and personally I would definitely welcome a PR.

Follow-up questions:

* To flesh out @vexx32 's concern: Since this would amount to effectively supporting namespace variable notation as the redirection target, we should probably support all drives:

  * Among the standard drives, aside from `variable:foo`, only `env:Foo` makes sense (writing to an environment variable), but conceivably there could be other providers whose drives could be meaningfully targeted.

* The original proposal included on additional feature variation; do we need it, and, if so, what should the syntax be?
  * Optionally also passing a stream _through_ while capturing it (this would be like an in-command `Tee-Object` for any stream targeted).

**Comment by KevinMarquette** [comment]
I would like it better as a cmdlet we can pipe to before we make it an operator.

**Comment by Jaykul** [comment]
I vote yes to making _all_ of the numbered streams generically redirectable to _any_ content provider PSDrive path

**Comment by SimonWahlin** [comment]
I really like the ability to redirect to a psdrive, please make it work on any content providers drive, not just variable. I guess variable and env are the ones that makes sense currently, but just as jaykul says, having it look like it is using a content provider when it is not will just create a special case that will be hard to explain.

**Comment by qaswani** [comment]
I love this. I re-implemented `Start-Process` to support this behavior so I could redirect `stderr`/`stdout` to a variable and return a `System.Diagnostics.Process` with two additional properties that contain the output. Having this feature would simplify a lot of code for parsing command output. It's only a matter of time before my hack runs into a scenario it can't handle.

**Comment by mklement0** [comment]
@Jaykul 

Agreed re `Tee-Object` cmdlet.

> Unless you're going to make interfaces public so that sort of host interception logging

I don't think we should think of this in terms of _hosts_ - the fact that external-program output is sent directly to the host by default should be considered an implementation detail. Instead, we want to be able to target _streams_:

The `>` operator already allows you to selectively redirect stderr output (which would by default go to the host, and with the proposal at hand we would now also support variables as targets (potentially among others).

The suggestion to enable optional _pass-through_ functionality stems from a gap in functionality with respect to external programs and their stderr output, specifically.

For _PowerShell_ commands you can use the `-*Variable` common parameters (e.g., `-ErrorVariable`), optionally via an aux. `Invoke-Command` call, to capture stream output _with pass-through logic_; that is, capturing stream output this way passes through _by default_, and you need to explicitly silence the target stream to suppress that.

This ability is currently lacking for _external-program calls_ - you only have the choice between passing through to the host only, or redirecting and thereby _invariably_ silencing the stream.

It is conceivable that you want to _both_ see stderr output - and doing so _in context_, interleaved with stdout output is especially important when calling external programs - _and_ capture it.

You can actually _somewhat_ emulate this with `Invoke-Command` at the moment though it's both highly obscure and has major side effects, notably polluting the success output stream and changing the output sequencing:

```powershell
# $e captures stderr output, while still passing it through - albeit invariably
# via the *success* output stream, and out of sequence
Invoke-Command { ls / nosuch 2>&1 } -ErrorVariable e
```

That said, if `-ErrorVariable` were able to capture stderr output _as-is_, then we arguably do _not_ need a new syntax (I cannot speak to the technical feasibility):

```powershell
# WISHFUL THINKING:
# Pass stderr output through, in output order, *and* capture it in variable $e
# (Currently, this records nothing in $e.)
Invoke-Command { ls / nosuch } -ErrorVariable e
```

---

> P.S. Alias: is a content drive and works just as well as ENV: for holding strings..

Indeed; to spell it out: the proposal is to allow targeting drives of all those PS providers that implement the `IContentCmdletProvider` interface, which _technically_ means the following standard providers:

```
PS> (Get-PSProvider | Where-Object { 
      $_.ImplementingType.GetInterfaces() -contains [System.Management.Automation.Provider.IContentCmdletProvider] 
    }).Name

Alias
Environment
FileSystem
Function
Variable

**Comment by Jaykul** [comment]
A new redirection syntax was also proposed for #5184 what ever happened to this idea @JamesWTruher?
